File: lib/core/models/notification_model.dart
class AppNotification {
  final String id;
  final String title;
  final String body;
  final String type;
  final bool isRead;
  final DateTime createdAt;
  final String? referenceId;

  const AppNotification({
    required this.id,
    required this.title,
    required this.body,
    required this.type,
    required this.isRead,
    required this.createdAt,
    this.referenceId,
  });

  factory AppNotification.fromMap(Map<String, dynamic> map) {
    return AppNotification(
      id: map['_id']?.toString() ?? map['id']?.toString() ?? '',
      title: map['title']?.toString() ?? '',
      body: map['body']?.toString() ?? '',
      type: map['type']?.toString() ?? 'system',
      isRead: map['isRead'] == true,
      createdAt: map['createdAt'] != null
          ? DateTime.parse(map['createdAt'].toString())
          : DateTime.now(),
      referenceId: map['referenceId']?.toString(),
    );
  }

  AppNotification copyWith({bool? isRead}) {
    return AppNotification(
      id: id,
      title: title,
      body: body,
      type: type,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt,
      referenceId: referenceId,
    );
  }
}

File: backend_temp/src/DB/models/notification.model.js
import mongoose, { Types } from "mongoose";

const schema = new mongoose.Schema(
  {
    user: {
      type: Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },
    title: {
      type: String,
      required: true,
      trim: true,
    },
    body: {
      type: String,
      required: true,
      trim: true,
    },
    type: {
      type: String,
      enum: ["transaction", "reminder", "offer", "system"],
      default: "system",
    },
    isRead: {
      type: Boolean,
      default: false,
    },
    referenceId: {
      type: String,
      default: null,
    },
  },
  {
    timestamps: true,
    versionKey: false,
  }
);

schema.index({ user: 1, createdAt: -1 });
schema.index({ user: 1, isRead: 1 });

schema.statics.findForUser = function (userId, limit = 50) {
  return this.find({ user: userId }).sort({ createdAt: -1 }).limit(limit);
};

export const Notification = mongoose.model("Notification", schema);
